rpcs3/Emu/RSX/Common/BufferUtils.cpp

```diff
@@ -342,111 +342,74 @@ void write_index_array_for_non_indexed_non_native_primitive_to_buffer(char* dst,
-{
-	u32 address = rsx::get_address(rsx::method_registers.index_array_address(), rsx::method_registers.index_array_location());
-	rsx::index_array_type type = rsx::method_registers.index_type();
-
-	u32 type_size = gsl::narrow<u32>(get_index_type_size(type));
-
-
-	EXPECTS(rsx::method_registers.vertex_data_base_index() == 0);
-
-	bool is_primitive_restart_enabled = rsx::method_registers.restart_index_enabled();
-	u32 primitive_restart_index = rsx::method_registers.restart_index();
 
-	// Disjoint first_counts ranges not supported atm
-	for (int i = 0; i < first_count_arguments.size() - 1; i++)
+namespace
+{
+	/**
+	* Get first index and index count from a draw indexed clause.
+	*/
+	std::tuple<u32, u32> get_first_count_from_draw_indexed_clause(const std::vector<std::pair<u32, u32>>& first_count_arguments)
 	{
-		const std::tuple<u32, u32> &range = first_count_arguments[i];
-		const std::tuple<u32, u32> &next_range = first_count_arguments[i + 1];
-		EXPECTS(std::get<0>(range) + std::get<1>(range) == std::get<0>(next_range));
+		u32 first = std::get<0>(first_count_arguments.front());
+		u32 count = std::get<0>(first_count_arguments.back()) + std::get<1>(first_count_arguments.back()) - first;
+		return std::make_tuple(first, count);
 	}
-	u32 first = std::get<0>(first_count_arguments.front());
-	u32 count = std::get<0>(first_count_arguments.back()) + std::get<1>(first_count_arguments.back()) - first;
-	auto ptr = vm::ps3::_ptr<const T>(address + first * type_size);
 
-	switch (draw_mode)
+
+	// TODO: Unify indexed and non indexed primitive expansion ?
+	template<typename T>
+	std::tuple<T, T> write_index_array_data_to_buffer_impl(gsl::span<T> dst,
+		gsl::span<const be_t<T>> src,
+		rsx::primitive_type draw_mode, bool restart_index_enabled, u32 restart_index, const std::vector<std::pair<u32, u32> > &first_count_arguments,
+		std::function<bool(rsx::primitive_type)> expands)
 	{
-	case rsx::primitive_type::points:
-	case rsx::primitive_type::lines:
-	case rsx::primitive_type::line_loop:
-	case rsx::primitive_type::line_strip:
-	case rsx::primitive_type::triangles:
-	case rsx::primitive_type::triangle_strip:
-	case rsx::primitive_type::quad_strip:
-		return upload_untouched<T>({ ptr, count }, dst, is_primitive_restart_enabled, primitive_restart_index);
-	case rsx::primitive_type::polygon:
-	case rsx::primitive_type::triangle_fan:
-		return expand_indexed_triangle_fan<T>({ ptr, count }, dst, is_primitive_restart_enabled, primitive_restart_index);
-	case rsx::primitive_type::quads:
-		return expand_indexed_quads<T>({ ptr, count }, dst, is_primitive_restart_enabled, primitive_restart_index);
-	}
+		u32 first;
+		u32 count;
+		std::tie(first, count) = get_first_count_from_draw_indexed_clause(first_count_arguments);
 
-	throw EXCEPTION("Unknown draw mode");
+		switch (draw_mode)
```

Comment:
Isn't it easier to just
```
if (!expands(mode)) upload_untouched; return;
switch(...)
```
That will allow the renderer to have more control over which path is taken imo